### ISSUE TITLE
Inference helper

### ISSUE THREAD

**Issue by lgvaz** [lgvaz/mantisshrimp#78]
## 🚀 Feature
**Is your feature request related to a problem? Please describe.**
A lot of boilerplate code is required for doing inference, the current necessary steps are currently described [here](https://www.kaggle.com/okeaditya/mantisshrimp-fasterrcnn#Making-Predictions-Work-in-Progress-Stay-tuned)

**Describe the solution you'd like**
Use a parser for extracting all the necessary information from a test set, and then calling a function that does the inference loop automatically.

Dream API:
```python
test_records = parser.parse()
test_dataset = Dataset(test_rs)
test_dataloader = model.dataloader(test_dataset)
model.inference(test_dataloader)
```

**Comment by lgvaz** [comment]
First proposal:

```python
class WheatTestParser(Parser, FilepathParserMixin):
    def __init__(self, img_dir):
        self.filepaths = get_image_files(img_dir)
        self.imageid_map = IDMap()

    def __iter__(self):
        yield from self.filepaths

    def __len__(self):
        return len(self.filepaths)

    def imageid(self, o) -> int:
        return o.stem

    def filepath(self, o) -> Union[str, Path]:
        return o


source = Path("/Users/lgvaz/.data/wheat")
df = pd.read_csv(source / "train.csv")

test_parser = WheatTestParser(source / "test")
test_records = test_parser.parse()[0]
test_dataset = Dataset(test_records)
```

Dataset eases out the process by opening the images, masks if required, and other steps to prepare out data. It also allows the usage of transforms if necessary

It's still left how should we create a dataloader, since the dataloader will always look the same for this model, I propose we implement a method `test_dataloader` on the model. We can see if lightning can help us with basics boring operations like switching the batch to GPU and stuff like that

@oke-aditya , what are your thoughts on this?

**Comment by oke-aditya** [comment]
Great prototype. This does what people do most times.

Also for each model, we need different test_dataloader as it might need different parameters. This should make stuff faster and efficient.

**Comment by lgvaz** [comment]
Perfect! I'm working on a PR and I'll keep you updated

**Comment by oke-aditya** [comment]
The above is good for Batch_inference. That is predicting from an Image Folder or in Kaggle Competitions.

Some cases we need to load a single image being uploaded by user, load model, run model, draw boxes, etc.

This too takes a lot of boiler plate.

**Comment by oke-aditya** [comment]
Dream API

```
model = create_model() # Let user make it using mantisshrimp we can't help here
model = load_model() # Load the model to device with state dicts, set to eval mode, etc.
image = get_image_tensor(image_path) # User can give image path or url or simply byteobjects as recieved in backend
# Also we can perform transforms that user needs to do.
prediction = model(image) # Simple pytorch code now.
```

- If the user gives us a well-formatted input then we can provide him bboxes for his image.
- If the prediction is a dictionary as 
```
{
   'boxes' :
   'labels' :
   'scores': # Confidence thershold    
}
```

We can draw the boxes, choose objects above the threshold, and do NMS as well.
This can be reused in the step of batch predictions too

For the user, it will be
```
image_with_boxes = processess_predictions(prediction, confidence_threshold, image_path, NMS threshold)
```

And he gets the simple image with boxes within a few steps. No worries how this stuff should be done.

for these ops, I propose a folder called inference.

So user can do ``` from mantisshrimp.infrence import * ```

**Comment by ai-fast-track** [comment]
Fastai has 2 methods for inference (prediction):

1- `[Learner.get_preds](https://github.com/fastai/fastai2/blob/bf455de9bc37c76f7f92b3c43227ef9d4779b614/fastai2/learner.py#L222)`: Prediction for a batch of items (images)
2- `[Learner.predict()](https://github.com/fastai/fastai2/blob/bf455de9bc37c76f7f92b3c43227ef9d4779b614/fastai2/learner.py#L246)` : Prediction for one item (image). 

Under the hood the `predict()` transform the item in a batch of one item, and calls `get_preds()`

**Comment by ai-fast-track** [comment]
Fastai has 2 methods for inference (prediction):

1- [Learner.get_preds](https://github.com/fastai/fastai2/blob/bf455de9bc37c76f7f92b3c43227ef9d4779b614/fastai2/learner.py#L222): Prediction for a batch of items (images)
2- [Learner.predict()](https://github.com/fastai/fastai2/blob/bf455de9bc37c76f7f92b3c43227ef9d4779b614/fastai2/learner.py#L246) : Prediction for one item (image).

Under the hood the predict() transform the item in a batch of one item, and calls get_preds()

**Comment by oke-aditya** [comment]
@ai-fast-track  I guess this is how forward pass performs (higly unsure correct me if I'm wrong), we leave that to the engine (torch / fastai / lightning) and handle the rest processing.

**Comment by ai-fast-track** [comment]
Yes we leave that to the corresponding engine (@lgvaz  can you please confirm?)

My understanding is that [adapted_fastai_learner](https://github.com/lgvaz/mantisshrimp/blob/a255e7c4872958933cc138ef81a4069d62a75260/mantisshrimp/engines/fastai/learner/adapted_fastai_learner.py#L9) returns a fastai Learner object and uses a converted fastai dataloaders object. Therefore, we can create a fastai test dataloaders object and pass it on to fastai Learner methods (`predict` and `get_preds`).

For those interested in Inference in fastai2, I wrote this [post](https://forums.fast.ai/t/fastai-v2-recipes-tips-and-tricks-wiki/64486/3?u=farid) in fastai forum:

**Comment by lgvaz** [comment]
I'm currently unsure how inference is going to work, it's good that we're already started talking about this, I think it's one of the hardest things to get right.

What I have in mind is that we are going to have something in our models that does not depend on any other library, maybe each model can have a `predict` method that takes a **batch** and return the prediction, maybe as a dictionary like Aditya already suggested. I would like this method to always return library standard types (e.g. `BBox`, `Mask`) whenever possible.

I'm a big fan of the golang philosophy of having only one way of doing things (and seems users of the language love that feature as well), a method that works on batches is all that we need, single item is just a specific case of it.

We can then link our implementation to `learn.get_preds` and `learn.predict`, but we're probably not going to use any of the fastai code for it, we just need to provide the fastai interface and use our methods.

**Comment by oke-aditya** [comment]
I guess let us keep a standard or something similar. If a model does not return something. Leave it as empty [] in keys.

If we reach a final point then we can write inference utils. 
```
{
   'boxes' :
   'labels' :
   'scores': # Confidence thershold    
}
```

**Comment by lgvaz** [comment]
That's the goal, the only suggestion is to not include the keys that the models does not produce instead of leaving an empty list

**Comment by oke-aditya** [comment]
- Can we have a dummy Mantiss model so that we can work with these simple inference scripts? It would also help in #107 #117 as we can use these functions here too.

- We can start making scripts assuming some common dictionary format.  (Can easily be converted to JSON too later).

- Scripts that we can add, NMS, Confidence Thresholding, IoU, etc.

**Comment by lgvaz** [comment]
Sure, I'll start working on that once #122 is done

In the past I've implemented something related, take a look at `MantisRCNN.predict`. What is left do to there is convert the model predictions to library standard objects.

e.g. the model will return a list of tensors for the bounding boxes coordinates. We have to convert those to a list of `BBox`.

**Comment by oke-aditya** [comment]
`model predictions to library standard objects.` Can you please write the library standard here.

**Comment by lgvaz** [comment]
What I mean by that is to use the objects inside `mantisshrimp.core`

**Comment by ai-fast-track** [comment]
Check out how they did here:

https://github.com/microsoft/computervision-recipes/blob/16d2caf2db7b484e9bbae8a333902f1ee98ed64f/utils_cv/detection/model.py#L587

They have:
1- [def predict()](https://github.com/microsoft/computervision-recipes/blob/16d2caf2db7b484e9bbae8a333902f1ee98ed64f/utils_cv/detection/model.py#L587)
2- [def predict_dl()](https://github.com/microsoft/computervision-recipes/blob/16d2caf2db7b484e9bbae8a333902f1ee98ed64f/utils_cv/detection/model.py#L626)
3-  [def predict_batch()](https://github.com/microsoft/computervision-recipes/blob/16d2caf2db7b484e9bbae8a333902f1ee98ed64f/utils_cv/detection/model.py#L641)

Notebook where they use those methods:
https://github.com/microsoft/computervision-recipes/blob/16d2caf2db7b484e9bbae8a333902f1ee98ed64f/scenarios/detection/01_training_introduction.ipynb

**Comment by oke-aditya** [comment]
Suggestion. 

Pytorch Lightning is one of our requirments. Why reinvent the wheel ? For inference we can simply use lightning metrics and get the job done. 

Otherwise use lot of code from [here](https://github.com/PyTorchLightning/pytorch-lightning/tree/master/pytorch_lightning/metrics)
